### PR TITLE
TSP API GET /shipments query params [ part of 159287411 ]

### PR DESCRIPTION
**Pull request by chrisgilmerproj**
## Description

The original query for `/shipments` ignored query parameters.  This updates the handler and model functions to handles them. 

## Reviewer Notes

Some changes to take note of:
- I was able to remove the raw SQL query that I'd started with in the previous PR and moved to using the built-in Pop functions.
- The `Paginate()` function obscures the use of `offset` in the api.  We could either build our own `Offset()` function and add it to Pop or I could attempt to add some raw SQL to the query to make this behave as expected.  For now I think we can just move on and revisit this later.
- The `status` query parameter is now an enum so that TSPs don't have to guess about different status types.  I'd like to connect it to the `ShipmentStatus` model but am unsure how to do it just now.
- I reordered some fields in structs to ensure they are readable and consistent across different parts of the code base.

For testing:
- There are now some more testing functions available for TSPs and Shipments.  Including `authenticateTspRequest`, `MakeDefaultShipmentOffer`, and `CreateShipmentOfferData`.  The latter is a factory for creating a set of shipments and shipment offers.  
- I renamed `pkg/testdatagen/make_offered_shipment_record.go` to `pkg/testdatagen/make_shipment_offer_record.go` to be more consistent with naming.
- The creation of ShipmentOffer objects now uses the `Assertions` model for setting and using defaults in tests.

## Code Review Verification Steps

* [x] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/blob/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* [x] Request review from a member of a different team.
* [x] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/n/projects/2181745/stories/159287411) for this change

### CODE REVIEW

File: swagger/api.yaml [hunk 1]
```diff
@@ -218,6 +218,18 @@ paths:
...
+            - COMPLETE
+            - AWARDED
+            - ACCEPTED
+            - REJECTED
+            - IN_TRANSIT
+            - SIT_ORIGIN
+            - SIT_DEST
+            - DELIVERED
+            - COMPLETED
+            - CLAIMS
```
Comment (chrisgilmerproj) [inline swagger/api.yaml]: I don't want them to have to guess what values work here.  Also wasn't quite sure how to reference the `ShipmentStatus` info directly via a `$ref`.

File: pkg/models/shipment.go [hunk 2]
```diff
@@ -116,23 +116,66 @@ func FetchShipments(dbConnection *pop.Connection, onlyUnassigned bool) ([]Shipme
...
+			query.Order(*orderBy)
+		}
+	}
+
+	// Manage limit and offset values
+	defaultLimit := int64(25)
+	if limit == nil {
+		limit = &defaultLimit
+	} else if *limit < int64(1) {
+		*limit = int64(1)
```
Comment (ntwyman) [inline pkg/models/shipment.go]: Do you intend to update someone else's variable? This will change the value that you have been passed a reference to.

On reflection I am not sure why these are taking references.

File: pkg/models/shipment.go [hunk 3]
```diff
@@ -116,23 +116,66 @@ func FetchShipments(dbConnection *pop.Connection, onlyUnassigned bool) ([]Shipme
 }
 
 // FetchShipmentsByTSP looks up all shipments belonging to a TSP ID
-func FetchShipmentsByTSP(tx *pop.Connection, tspID uuid.UUID) ([]Shipment, error) {
-	shipments := []Shipment{}
-
-	var sql string
+func FetchShipmentsByTSP(tx *pop.Connection, tspID uuid.UUID, status *string, orderBy *string, limit *int64, offset *int64) ([]Shipment, error) {
```
Comment (ntwyman) [inline pkg/models/shipment.go]: See comment below. I think the limit and offset should be passed by value rather tan reference.

File: swagger/api.yaml (hunk 1 above)
Comment (chrisgilmerproj) [reply swagger/api.yaml]: So I tried but the `$ref` brings in a bunch of fields from that definition that aren't allowed here.  I had considered creating another definition which was just `ShipmentStatusEnum` to use in both places.  What do you think?

File: pkg/models/shipment.go (hunk 2 above)
Comment (chrisgilmerproj) [reply pkg/models/shipment.go]: I had actually meant to modify the original struct but if that's not a pattern we like I can pass by value.  The original reason I passed the reference was so that all the validation happened only in this models file instead of splitting validation between the handlers and models files.  

The other reason is that when you pass a default as `Nil` from swagger the struct that is returned has not set that property in the struct.  I was surprised that it did that instead of just setting nullable fields to `Nil` but perhaps its to save memory overhead. 

So two questions:
- Should I do validation in the handler or the model file? Or split between them (see the sorting logic)
- Should I create new variables to pass along to the models file?

File: pkg/models/shipment.go (hunk 3 above)
Comment (chrisgilmerproj) [reply pkg/models/shipment.go]: Do you mean that I should pass only some of them as by-value instead of by-reference? Like passing mixed types?

File: swagger/api.yaml (hunk 1 above)
Comment (amitch23) [reply swagger/api.yaml]: It looks like the enum values you define here are the same as are in '#/definitions/ShipmentStatus'  - what other fields are brought in? What would `ShipmentStatusEnum` contain that `ShipmentStatus` doesn't?

Comment (chrisgilmerproj) [reply swagger/api.yaml]: So I tried this syntax in the `parameters` section:

```
        - name: status
          in: query
          $ref: '#/definitions/ShipmentStatus'
```

And got this error:

```
The swagger spec at "/Users/cgilmer/Projects/go/src/github.com/transcom/mymove/swagger/api.yaml" is invalid against swagger specification 2.0. see errors :
- "paths./shipments.get.parameters" must validate one and only one schema (oneOf). Found none valid
- paths./shipments.get.parameters.$ref in body is a forbidden property
- paths./shipments.get.parameters.in in body should be one of [header]
- paths./shipments.get.parameters.type in body is required
```

Comment (chrisgilmerproj) [reply swagger/api.yaml]: @amitch23 had me try this:  

```
        - name: status
          in: query
          schema:
            $ref: '#/definitions/ShipmentStatus'
```

and got these errors:

```
The swagger spec at "/Users/cgilmer/Projects/go/src/github.com/transcom/mymove/swagger/api.yaml" is invalid against swagger specification 2.0. see errors :
- "paths./shipments.get.parameters" must validate one and only one schema (oneOf). Found none valid
- paths./shipments.get.parameters.in in body should be one of [body]
```

Because the one difference is `in: query`.

Comment (chrisgilmerproj) [reply swagger/api.yaml]: I used the example from here to fix this: https://swagger.io/docs/specification/2-0/enums/

tl;dr - YAML references saved the day.  However, it required that I put the `definitions` section ahead of the `paths` section, which is what had been done in the `swagger/internal.yaml` anyway, so it seemed like an appropriate change.

**Comment by chrisgilmerproj** [comment]
@ntwyman and @amitch23 - I put in a few changes here.  Is there anything else I should do here before approval?
